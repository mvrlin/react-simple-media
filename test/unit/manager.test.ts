import { MediaQueryManager } from '~/manager'

describe('MediaQueryManager', () => {
  beforeEach(() => {
    // @ts-ignore
    MediaQueryManager._instance = undefined
  })

  it('should use the same instance', () => {
    const managerA = new MediaQueryManager()
    const managerB = new MediaQueryManager()

    expect(managerA === managerB).toBeTruthy()
  })

  describe('getMediaQueries', () => {
    it('should generate and cache queries', () => {
      const manager = new MediaQueryManager()

      expect(manager.getMediaQueries()).toBe(manager.getMediaQueries())
      expect(manager.getMediaQueries()).toMatchSnapshot()
    })
  })

  describe('usePreset', () => {
    it('should use presets', () => {
      const manager = new MediaQueryManager()
      manager.usePreset('tailwind')

      expect(manager.getMediaQueries()).toMatchSnapshot()
    })
  })

  describe('detectBreakpoint', () => {
    it('should detect breakpoint by cookie', async () => {
      const cookieName = 'mq'

      const manager = new MediaQueryManager({ cookieName }).usePreset('tailwind')
      const breakpoint = await manager.detectBreakpoint(`${cookieName}=xs`)

      expect(breakpoint).toBe('xs')
    })

    it('should detect breakpoint by user-agent', async () => {
      const userAgent =
        'Mozilla/5.0 (iPad; CPU OS 15_5 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) CriOS/103.0.5060.63 Mobile/15E148 Safari/604.1'

      const manager = new MediaQueryManager().usePreset('tailwind')
      const breakpoint = await manager.detectBreakpoint('', userAgent)

      expect(breakpoint).toBe('md')
    })

    it('should fallback to default breakpoint', async () => {
      const fallbackBreakpoint = 'lg'

      const manager = new MediaQueryManager({ fallbackBreakpoint }).usePreset('tailwind')
      const breakpoint = await manager.detectBreakpoint()

      expect(breakpoint).toBe(fallbackBreakpoint)
    })
  })
})
