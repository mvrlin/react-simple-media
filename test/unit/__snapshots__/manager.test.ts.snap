// Vitest Snapshot v1

exports[`MediaQueryManager > getMediaQueries > should generate and cache queries 1`] = `
{
  "desktop": "(min-width: 1024px) and (max-width: 1279px)",
  "desktopMedium": "(min-width: 1280px) and (max-width: 1599px)",
  "desktopWide": "(min-width: 1600px)",
  "mobile": "(min-width: 1px) and (max-width: 374px)",
  "mobileMedium": "(min-width: 375px) and (max-width: 424px)",
  "mobileWide": "(min-width: 425px) and (max-width: 767px)",
  "tablet": "(min-width: 768px) and (max-width: 1023px)",
}
`;

exports[`MediaQueryManager > usePreset > should use presets 1`] = `
{
  "2xl": "(min-width: 1536px)",
  "lg": "(min-width: 1024px) and (max-width: 1279px)",
  "md": "(min-width: 768px) and (max-width: 1023px)",
  "sm": "(min-width: 640px) and (max-width: 767px)",
  "xl": "(min-width: 1280px) and (max-width: 1535px)",
  "xs": "(min-width: 1px) and (max-width: 639px)",
}
`;
